trie/node.go

```diff
@@ -165,45 +174,58 @@ func (l *leaf) Encode() ([]byte, error) {
 	return encoding, nil
 }
 
-func (b *branch) header() []byte {
+func (b *branch) header() ([]byte, error) {
 	var header byte
 	if b.value == nil {
 		header = 2 << 6
 	} else {
 		header = 3 << 6
 	}
 	var encodePkLen []byte
+	var err error
 
 	if len(b.key) >= 63 {
 		header = header | 0x3f
-		encodePkLen = encodeExtraPartialKeyLength(len(b.key))
+		encodePkLen, err = encodeExtraPartialKeyLength(len(b.key))
+		if err != nil {
+			return nil, err
+		}
 	} else {
 		header = header | byte(len(b.key))
 	}
 
 	fullHeader := append([]byte{header}, encodePkLen...)
-	return fullHeader
+	return fullHeader, nil
 }
 
-func (l *leaf) header() []byte {
+func (l *leaf) header() ([]byte, error) {
 	var header byte = 1 << 6
 	var encodePkLen []byte
+	var err error
 
 	if len(l.key) >= 63 {
 		header = header | 0x3f
-		encodePkLen = encodeExtraPartialKeyLength(len(l.key))
+		encodePkLen, err = encodeExtraPartialKeyLength(len(l.key))
+		if err != nil {
+			return nil, err
+		}
 	} else {
 		header = header | byte(len(l.key))
 	}
 
 	fullHeader := append([]byte{header}, encodePkLen...)
-	return fullHeader
+	return fullHeader, nil
 }
 
-func encodeExtraPartialKeyLength(pkLen int) []byte {
+func encodeExtraPartialKeyLength(pkLen int) ([]byte, error) {
 	pkLen -= 63
 	fullHeader := []byte{}
-	for i := 0; i < 317; i++ {
+
+	if pkLen >= 1<<16 {
+		return nil, errors.New("partial key length exceeds 2^16")
```

Comment:
This may be nitpicking but maybe we should say:
"partial key length is greater than or equal to 2^16"
or
"partial key length is 2^16 or greater"

I'll leave it to you though. Don't necessarily have to change anything.

Reply:
I changed it to make it clearer

Reply:
Awesome!